### DISCUSSION TITLE
Pyodide / MicroPython `version` in their config ???

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1759] [Proposals]
With *polyscript* each script tag could have its own specific version out of an attribute, but this is not, and likely should never be, the case for *PyScript*, as we all know multiple config and multiple things mean multiple pyodide too + we can't guarantee anymore stuff "*just works*" because our code is developed and tested on specific versions of each interpreter ... **however** ... from a developer perspective, instead of having the `interpreters` array in `py-config` that never actually worked as meant because we inevitably run only one interpreter, I'd rather suggest a `version` field that would pass through *polyscript* and use whatever explicit version it is ... reasons:

  * it might be handy for testing purpose for both us and users / contributors ... a new version of Pyodide or MicroPython is out and we all would like to see if it breaks our App *before* updating to the latest/greatest (as we also recommend pinning versions of the library)
  * we can use it to eventually debug regressions out of interpreters we use
  * interpreters developers themselves can use fully qualified URLs to point at their local branch for bug fixes and tests

That's a lot of help for everyone to move forward with ease so that the proposal is:

```html
<py-config>
version = "0.23.4"
</py-config>
<mpy-config>
version = "0.1.239.X"
</mpy-config>
```

both options are passed along and we have easy life testing all we want from anywhere we like ... thoughts?

/cc @antocuni @ntoll @JeffersGlass @bugzpodder

**Comment by WebReflection** [comment]
P.S. if `version` is too generic, `interpreter = "x.x.x"` would work to me as well, or any other name you like.

**Comment by JeffersGlass** [comment]
I think this is quite good! Just to expand on your point about fully qualified URLs, does that mean something like this would also work?

```html
<py-config>
version = "http://127.0.0.1:5502/local/build/of/pydodide.js"
</py-config>
```

As for naming... I do like `interpreter` better than `version`, just cause version in py-config sounds like a "PyScript" version or something, which it surely is not.

**Reply by WebReflection** [reply]
Yes, fully qualified URLs are already handled well in *polyscript* and that is indeed a perfect use case for using them in development / debugging sessions.

**Comment by antocuni** [comment]
I think this is a good idea, as long as we make sure that it's a "best effort" option.

In pyscript classic, we had cases in which pyodide's internal API changed and we had complicated logic to try to work across versions, but it was not tested and the logic tended to stay there forever. I'd like to avoid that.

As for the syntax, what about the following?
```
<py-config>
pyodide = "0.24.1"
#but also: pyodide = "https://..../pyodide.js"
</py-config>

<mpy-config>
micropython = "1.2.3"
</mpy-config>
```

**Reply by WebReflection** [reply]
that couples the interpreter name to the tag ... we discussed the possibility to change `micropython` to `micropython-wasm` as example, and that would break every code to date that used `micropython` so it's a future hostile choice to me, as we already have a single interpreter per tag, hence it's also a bit redundant.

If we go for `interpreter` instead, it really doesn't matter how that's called and if we want to have `<ruby-script>` or others in the future, nobody would need to remember `ruby-wasm-wasi = "...."` and every `<ruby-config>` can use `interpreter` there.

The only *real* issue to me is the fact two different custom types cannot share the same `.toml` or `.json` file as `config` attribute, but as this feature is mostly meant for developers and debugging sessions + the two custom types really have little in common to share beside eventually `fetch`, I don't think that's a compelling argument to go for something more complicated and/or problematic that binds an interpreter name to a tag + I don't want to deal with "*you wrote micropython in py-config*" errors as we said anyone can use whatever key they want in there so that reserving just `interpreter` feels also easier to reason about and less to worry, imho.

**Comment by antocuni** [comment]
> that couples the interpreter name to the tag ... we discussed the possibility to change `micropython` to `micropython-wasm` as example, and that would break every code to date that used `micropython` so it's a future hostile choice to me, as we already have a single interpreter per tag, hence it's also a bit redundant.

right

> If we go for `interpreter` instead, it really doesn't matter how that's called and if we want to have `<ruby-script>` or others in the future, nobody would need to remember `ruby-wasm-wasi = "...."` and every `<ruby-config>` can use `interpreter` there.

ok, you convinced me :)

**Comment by ntoll** [comment]
Catching up with GitHub after my Wednesday "hermit mode".

This all sounds like a good idea to me. I definitely prefer `interpreter` to `version` or an explicit name of an interpreter to indicate the interpreter's version number. :+1:

**Reply by WebReflection** [reply]
the explicit name has been nuked as option https://github.com/pyscript/pyscript/discussions/1759#discussioncomment-7120485 but we all like `interpreter` so let it be `interpreter` :partying_face:
